edx_notifications/base_data.py

```diff
@@ -0,0 +1,410 @@
+        """
+
+        if not self._expected_types:
+            raise TypeError(
+                "Subclass of TypedField ('{type_name}') has not set required "
+                "_expected_types attribute".format(type_name=type(self).__name__)
+            )
+
+        self._default = kwargs.get('default', None)
+
+    def _assert_has_name(self):
+        """
+        Helper method to make sure we have our __name__ attribute
+        set before we do any reads/writes
+        """
+
+    def __get__(self, data_object, owner_class):
+        """
+        Descriptor getter
+        """
+        if not data_object:
+            # called as a class method, so return the descriptor (ourselves)
+            return self
+
+        self._assert_has_name()
+
+        if not hasattr(data_object, '_field_data'):
+            return self._default
+
+        return data_object._field_data.get(self.__name__, self._default)  # pylint: disable=protected-access
+
+    def __set__(self, data_object, value):
+        """
+        Descriptor setter. Be sure to enforce type on setting. But None is allowed.
+        """
+
+        self._assert_has_name()
+
+        value_type = type(value)
+
+        if value and value_type not in self._expected_types:
+            raise TypeError(
+                (
+                    "Field expected type of '{expected}' got '{got}'"
+                ).format(expected=self._expected_types, got=value_type)
+            )
+
+        if not hasattr(data_object, '_field_data'):
+            data_object._field_data = {}  # pylint: disable=protected-access
+
+        data_object._field_data[self.__name__] = value  # pylint: disable=protected-access
+
+    def __delete__(self, data_object):
+        """
+        Descriptor delete
+        """
+
+        self._assert_has_name()
+
+        if hasattr(data_object, '_field_data') and self.__name__ in data_object._field_data:  # pylint: disable=protected-access
+            del data_object._field_data[self.__name__]  # pylint: disable=protected-access
+
+
+class StringField(TypedField):
+    """
+    Specialized subclass of TypedField(unicode) as a convienence
+    """
+
+    _expected_types = [unicode, str]
+
+
+class IntegerField(TypedField):
+    """
+    Specialized subclass of TypedField(int) as a convienence
+    """
+
+    _expected_types = [int, long]
+
+
+class DictField(TypedField):
+    """
+    Specialized subclass of TypedField(dict) as a convienence
+    """
+
+    _expected_types = [dict]
+
+    @classmethod
+    def to_json(cls, data):
+        """
+        Serialize to json
+        """
+
+        if not data:
+            return None
+
+        def datetime_to_json(obj):
+            """
+            JSON serializer for objects not serializable by default json code.
+            For now, this means datetime objects.
+            """
+
+            if isinstance(obj, datetime):
+                serial = obj.isoformat()
+                return serial
+
+            return obj
```

Comment:
This doesn't behave properly: if it can't serialize the object, it should raise `TypeError` (the same way that the default argument to `default` does https://docs.python.org/2/library/json.html#json.dump).